### PR TITLE
Windows store banner

### PR DESCRIPTION
**Pull request by cait1inhennessy**
capture
capture2

Goes with https://github.com/Microsoft/pxt-adafruit/pull/483

- Generic class for notification banners
- Options to specify delay before banner pops out, duration of display, and color theme
- Used to create windows store banner and experimental banner (pictured above)

### CODE REVIEW

File: webapp/src/app.tsx [hunk 1]
```diff
@@ -1755,7 +1767,29 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
 
         return (
             <div id='root' className={rootClasses}>
-                {showExperimentalBanner ? <container.ExperimentalBanner parent={this} /> : undefined}
+                {showExperimentalBanner ? <div id="experimentalBanner" className="ui icon top attached fixed negative mini message">
```
Comment (pelikhan) [inline webapp/src/app.tsx]: We should move this into a separate component to keep app.tsx small

File: theme/common.less [hunk 2]
```diff
@@ -362,10 +415,103 @@ div.simframe > iframe {
...
+
+#windowsStoreBanner img {
+    height: calc(@bannerHeight - 1rem);
+}
+
+/*******************************
+             Docs
+*******************************/
+
+#docs {
```
Comment (pelikhan) [inline theme/common.less]: Are these CSS changes related to the Windows banner? We should have a separate .less file for it.

File: webapp/src/app.tsx [hunk 3]
```diff
@@ -1755,7 +1767,29 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
...
+                    <sui.Icon icon="close" onClick={() => this.hideBanner()} />
+                    <div className="content">
+                        <div className="header">{lf("You are viewing an experimental version of the editor")}</div>
+                        <a href={liveUrl}>{lf("Take me back")}</a>
+                    </div>
+                </div> : undefined}
+                {showWindowsStoreBanner ? <div id="windowsStoreBanner" className="ui blue attached message">
+                    <div className="thumbnail">
+                        <a href={pxt.appTarget.appTheme.windowsStoreLink}>
+                            <img src="https://assets.windowsphone.com/13484911-a6ab-4170-8b7e-795c1e8b4165/English_get_L_InvariantCulture_Default.png" alt="Get" />
```
Comment (pelikhan) [inline webapp/src/app.tsx]: Store this image in our cloud.

File: webapp/src/app.tsx [hunk 4]
```diff
@@ -1755,7 +1767,29 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
...
+                    <sui.Icon icon="warning circle" />
+                    <sui.Icon icon="close" onClick={() => this.hideBanner()} />
+                    <div className="content">
+                        <div className="header">{lf("You are viewing an experimental version of the editor")}</div>
+                        <a href={liveUrl}>{lf("Take me back")}</a>
+                    </div>
+                </div> : undefined}
+                {showWindowsStoreBanner ? <div id="windowsStoreBanner" className="ui blue attached message">
+                    <div className="thumbnail">
+                        <a href={pxt.appTarget.appTheme.windowsStoreLink}>
```
Comment (pelikhan) [inline webapp/src/app.tsx]: Missing analytics

File: webapp/src/app.tsx [hunk 5]
```diff
@@ -1755,7 +1767,29 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
...
+                            <img src="https://assets.windowsphone.com/13484911-a6ab-4170-8b7e-795c1e8b4165/English_get_L_InvariantCulture_Default.png" alt="Get" />
+                        </a>
+                    </div>
+                    <div className="text">
+                        <a href={pxt.appTarget.appTheme.windowsStoreLink}>
+                            Get the app from the <span className="windowsStore">Windows Store</span>
```
Comment (pelikhan) [inline webapp/src/app.tsx]: Use lf for localization

File: pxteditor/editor.ts [hunk 6]
```diff
@@ -40,6 +40,7 @@ namespace pxt.editor {
         tutorialOptions?: TutorialOptions;
 
         running?: boolean;
+        home?: boolean;
```
Comment (pelikhan) [inline pxteditor/editor.ts]: why moving this?

File: pxteditor/editor.ts [hunk 7]
```diff
@@ -54,8 +55,8 @@ namespace pxt.editor {
         tracing?: boolean;
 
         highContrast?: boolean;
-
-        home?: boolean;
+        hideExperimentalBanner?: boolean;
+        hideWindowsStoreBanner?: boolean;
```
Comment (pelikhan) [inline pxteditor/editor.ts]: the default state should be no experimental banner, so this state should be reversed.

File: webapp/src/app.tsx [hunk 8]
```diff
@@ -1755,10 +1767,30 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
...
                     <header className="menubar" role="banner">
                         {inEditor ? <accessibility.EditorAccessibilityMenu parent={this} highContrast={this.state.highContrast}/> : undefined }
+
+                        {showWindowsStoreBanner ? <div id="windowsStoreBanner" className="ui attached message">
+                            <sui.Link class="link" target="_blank" ariaLabel={lf("View app in the Windows Store")} href={pxt.appTarget.appTheme.windowsStoreLink}>
+                                <span>
+                                    <img src="https://assets.windowsphone.com/13484911-a6ab-4170-8b7e-795c1e8b4165/English_get_L_InvariantCulture_Default.png"
+                                    alt={lf("Windows Store logo")} />
+                                </span>
+                                {lf("Get the app from the")} <span className="bold">{lf("Windows Store")}</span>
```
Comment (pelikhan) [inline webapp/src/app.tsx]: Composing strings does not translate well for languages that have different word ordering. It has to be a single strings in a lf. For simplicity drop the bold.

**Comment by microsoftsam** [comment]
Possibly also a merge conflict at some point.

File: pxteditor/editor.ts [hunk 9]
```diff
@@ -56,6 +56,7 @@ namespace pxt.editor {
         highContrast?: boolean;
 
         home?: boolean;
+        hideWindowsStoreBanner?: boolean;
```
Comment (microsoftsam) [inline pxteditor/editor.ts]: Not needed anymore?

File: theme/common.less [hunk 10]
```diff
@@ -361,11 +355,75 @@ div.simframe > iframe {
...
+        line-height: @bannerHeight;
+        cursor: pointer;
+        &:hover, &:focus {
+            outline: none;
+            color: darken(@bannerContrastColor, 25%);
+        }
+    }
+}
+
+#windowsStoreBanner
```
Comment (microsoftsam) [inline theme/common.less]: No need to move this. Keep this in the Layout section. (at the top)

Comment (microsoftsam) [reply theme/common.less]: I'd almost go as far as just removing it from this list and adding a top: 0 in it's own #windowStoreBanner style definition. This bit of CSS is for hiding the menu bar.

File: theme/themes/pxt/globals/site.variables [hunk 11]
```diff
@@ -231,4 +231,11 @@
 --------------------*/
 
 @tutorialSegmentBackground: #f8f8f9;
-@tutorialSegmentColor: @textColor;
\ No newline at end of file
+@tutorialSegmentColor: @textColor;
+/*-------------------
```
Comment (microsoftsam) [inline theme/themes/pxt/globals/site.variables]: nit: new line

File: webapp/src/app.tsx [hunk 12]
```diff
@@ -1726,6 +1731,9 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
         const isExperimentalUrlPath = location.pathname !== "/"
             && (targetTheme.appPathNames || []).indexOf(location.pathname) === -1;
         const showExperimentalBanner = !isLocalServe && isApp && isExperimentalUrlPath;
+        const isWindows10 = pxt.BrowserUtils.isWindows10();
+        const showWindowsStoreBanner = !this.state.hideWindowsStoreBanner && !pxt.winrt.isWinRT() && isWindows10 && pxt.appTarget.appTheme.windowsStoreLink;
+        const liveUrl = pxt.appTarget.appTheme.homeUrl + location.search + location.hash;
```
Comment (microsoftsam) [inline webapp/src/app.tsx]: This is a merge conflict. liveUrl is not used here.

Comment (cait1inhennessy) [reply webapp/src/app.tsx]: Fixed

File: theme/themes/pxt/globals/site.variables (hunk 11 above)
Comment (cait1inhennessy) [reply theme/themes/pxt/globals/site.variables]: Fixed

File: theme/common.less (hunk 10 above)
Comment (cait1inhennessy) [reply theme/common.less]: Don't understand the request - you want me to move this one rule back to the Layout section?  Or all windowsStoreBanner rules?

File: webapp/src/container.tsx [hunk 13]
```diff
@@ -443,6 +443,31 @@ export class CookieMessage extends data.Component<CookieMessageProps, CookieMess
     }
 }
 
+export class WindowsStoreBanner extends data.Component<ISettingsProps, {}> {
```
Comment (pelikhan) [inline webapp/src/container.tsx]: Move to its own container.

**Comment by cait1inhennessy** [comment]
@pelikhan @microsoftsam Ping for review of new code

File: webapp/src/notification.tsx [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    //TODO need shouldComponentUpdate?
+
+    constructor(props: NotificationBannerProps) {
+        super(props);
+        this.delayTime = this.props.delayTime || 0;
+        this.doneSleeping = this.sleepDone();
+        this.bannerType = this.props.bannerType || "default";
+        this.state = {visible: false};
+
+        setTimeout(() => this.show(), this.delayTime);
```
Comment (pelikhan) [inline webapp/src/notification.tsx]: I think we should move this logic into a initialization function that runs in the app.tsx init phase. The constructor of a view should not be state changing.

File: webapp/src/app.tsx [hunk 15]
```diff
@@ -1726,6 +1735,8 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
         const isExperimentalUrlPath = location.pathname !== "/"
             && (targetTheme.appPathNames || []).indexOf(location.pathname) === -1;
         const showExperimentalBanner = !isLocalServe && isApp && isExperimentalUrlPath;
+        const isWindows10 = pxt.BrowserUtils.isWindows10();
+        const showWindowsStoreBanner = !pxt.winrt.isWinRT() && isWindows10 && Cloud.isOnline() && pxt.appTarget.appTheme.windowsStoreLink && !showExperimentalBanner;
```
Comment (pelikhan) [inline webapp/src/app.tsx]: should this logic be hidden in the notification module?

File: webapp/src/app.tsx [hunk 16]
```diff
@@ -1755,10 +1767,11 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
 
         return (
             <div id='root' className={rootClasses}>
-                {showExperimentalBanner ? <container.ExperimentalBanner parent={this} /> : undefined}
                 {hideMenuBar ? undefined :
                     <header className="menubar" role="banner">
                         {inEditor ? <accessibility.EditorAccessibilityMenu parent={this} highContrast={this.state.highContrast}/> : undefined }
+                        {showExperimentalBanner ? <notification.ExperimentalBanner parent={this} /> : undefined}
```
Comment (pelikhan) [inline webapp/src/app.tsx]: this one too could be hidden into a "Notification" component thayt takes care of the logic of showing/hiding

File: webapp/src/app.tsx [hunk 17]
```diff
@@ -1770,8 +1754,7 @@ ${compileService && compileService.githubCorePackage && compileService.gittag ?
                 {hideMenuBar ? undefined :
                     <header className="menubar" role="banner">
                         {inEditor ? <accessibility.EditorAccessibilityMenu parent={this} highContrast={this.state.highContrast}/> : undefined }
-                        {showExperimentalBanner ? <notification.ExperimentalBanner parent={this} /> : undefined}
-                        {showWindowsStoreBanner ? <notification.WindowsStoreBanner parent={this} /> : undefined}
+                        {<notification.NotificationBanner parent={this} ref="notificationBanner" />}
```
Comment (pelikhan) [inline webapp/src/app.tsx]: Curliest not needed

**Comment by cait1inhennessy** [comment]
@pelikhan @microsoftsam Ping for another review.  @pelikhan all initialization and display logic is now handled by the banner component; this should address your previous comments.

File: webapp/src/notification.tsx [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+        super(props);
+        this.delayTime = this.props.delayTime || 0;
+        this.doneSleeping = this.sleepDone();
+        this.bannerType = this.props.bannerType || "default";
+    }
+
+    componentDidMount() {
+        if (this.doneSleeping) {
+            setTimeout(() => this.show(), this.delayTime);
+            if (this.props.displayTime) {
```
Comment (pelikhan) [inline webapp/src/notification.tsx]: this code should be triggered on show?

File: pxteditor/editor.ts [hunk 19]
```diff
@@ -171,6 +172,9 @@ namespace pxt.editor {
         openInstructions(): void;
         closeFlyout(): void;
 
+        showBanner(): void;
```
Comment (microsoftsam) [inline pxteditor/editor.ts]: Could this be one function accepting a bool. Just to reduce the number of functions specific to this component in apptsx

File: theme/common.less [hunk 20]
```diff
@@ -27,13 +27,7 @@ html, body {
 }
 
 #filelist, #maineditor, #sidedocs {
-    top:@mainMenuHeight;
-}
-
-.hideMenuBar #filelist,
```
Comment (microsoftsam) [inline theme/common.less]: Keep these up here for the .hideMenuBar feature. 
Add top:0 to the notificationBanner CSS. 

we have CSS optimizers in place so it shouldn't matter.
